### PR TITLE
Fix problem with isolation mode when duplicate names occurr

### LINKED ISSUE #2915: Unexpected execution order and failure with duplicated test names
I tried with Kotest 5.0 and later updates

The issue seems to be related to be triggered by two tests _not on the same level_ having the same name. The following tests should work: each `load()` is properly followed by a `complete()`, and the tests pass on Kotest 4.x. However, on 5.x they fail with two exceptions: one somewhere in JUnit code, the other in the `FakeService`. Looks like either `InstancePerLeaf` is not respected, or the execution order is just messed up because of the duplicated names.

Renaming one of the duplicated test names causes the tests to pass, also removing some `it`s (I think).

This behavior is unexpected, because the "duplicated" test names appear on different levels. My understanding was that the names are considered duplicate on the same tests level only.

Repro:

```kotlin
package com.example
​
import io.kotest.core.TestConfiguration
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.DescribeSpec
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.TestCoroutineScope
​
fun TestConfiguration.customTestScope() = TestCoroutineScope(Job()).also { testScope ->
    afterSpec {
        testScope.cancel()
        testScope.cleanupTestCoroutines()
    }
}
​
internal class Repro : DescribeSpec() {
​
    override fun isolationMode() = IsolationMode.InstancePerLeaf
​
    private val service = FakeService()
​
    init {
        describe("test") {
            val load: suspend () -> Unit = { runCatching { service.await() } }
​
            describe("foo") {
                customTestScope().launch { load() }
​
                it("a") {}
​
                it("b") {}
​
                describe("THIS TEST NAME IS DUPLICATED") {
                    service.complete()
​
                    it("a") {}
​
                    it("b") {}
                }
​
                describe("bar") {
                    service.complete()
​
                    it("a") {}
​
                    it("b") {}
​
                    describe("baz") {
                        customTestScope().launch { load() }
​
                        it("a") {}
​
                        it("b") {}
​
                        describe("THIS TEST NAME IS DUPLICATED") {
                            service.complete()
​
                            it("a") {}
​
                            it("b") {}
                        }
                    }
                }
            }
        }
    }
}
​
class FakeService {
​
    var deferred: CompletableDeferred<Unit>? = null
​
    suspend fun await() {
        if (deferred != null) error("Can't await without completing first")
​
        val cd = CompletableDeferred<Unit>()
        deferred = cd
​
        val value = runCatching { cd.await() }
        deferred = null
        value.getOrThrow()
    }
​
    fun complete() {
        deferred!!.complete(Unit)
    }
}
```

### PR DESCRIPTION
**Pull request by Kantis**
Fixes #2915

### CODE REVIEW

File: kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/descriptors/descriptor.kt [hunk 1]
```diff
@@ -103,15 +103,15 @@ sealed interface Descriptor {
...
+      is TestDescriptor -> this == descriptor.parent
    }
 
    /**
     * Returns true if this descriptor is ancestor (1..nth-parent) of the given [descriptor].
     */
    fun isAncestorOf(descriptor: Descriptor): Boolean = when (descriptor) {
       is SpecDescriptor -> false // nothing can be an ancestor of a spec
-      is TestDescriptor -> this.id == descriptor.parent.id || isAncestorOf(descriptor.parent)
+      is TestDescriptor -> this == descriptor.parent || isAncestorOf(descriptor.parent)
```
Comment (sksamuel) [inline kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/descriptors/descriptor.kt]: could replace this with isParent || isAncestor

File: kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/descriptors/descriptor.kt [hunk 2]
```diff
@@ -103,15 +103,15 @@ sealed interface Descriptor {
     */
    fun isParentOf(descriptor: Descriptor): Boolean = when (descriptor) {
       is SpecDescriptor -> false // nothing can be the parent of a spec
-      is TestDescriptor -> this.id == descriptor.parent.id
+      is TestDescriptor -> this == descriptor.parent
```
Comment (sksamuel) [inline kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/descriptors/descriptor.kt:106]: This fails if two things on the same path have a parent name that is the same ?
Can we add an explicit test for this to the descriptor tests ?

Comment (Kantis) [reply kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/descriptors/descriptor.kt:106]: [Tests added](https://github.com/kotest/kotest/pull/3071/commits/b0230f0c43f29169b77bd33535b081e56104ed53)

**Comment by sksamuel** [comment]
Tests failed but otherwise lgtm.

Could add to release branch too for a 5.3.3 at some point

**Comment by Kantis** [comment]
@sksamuel needs an approval for merging :)

What's the plan for 5.4? Anything major we want to get done before pushing it?

**Review by sksamuel — APPROVED** [review]

**Pull request merged**